File: ios/ElderlyAssistant/Services/MedicationScheduler/PhotoVerifier.swift
import Foundation

// MARK: - Photo Verifier (FR-D04a through FR-D04e)

enum PhotoCaptureResult {
    case captured(imageData: Data)
    case cameraUnavailable(reason: String)
    case captureFailed(reason: String)
}

enum PhotoDeliveryResult {
    case delivered(deliveredAt: Date)
    case deliveryFailed(reason: String)
}

protocol PhotoCaptureSession {
    func captureStillImage() async -> PhotoCaptureResult
}

protocol PhotoDeliveryChannel {
    func deliver(photo: VerificationPhoto, to targetDeviceId: String) async -> PhotoDeliveryResult
}

final class PhotoVerifier {
    private let captureSession: PhotoCaptureSession
    private let deliveryChannel: PhotoDeliveryChannel
    private let maxImageSizeBytes: Int = 200 * 1024   // 200 KB per FR-D04b

    init(captureSession: PhotoCaptureSession, deliveryChannel: PhotoDeliveryChannel) {
        self.captureSession = captureSession
        self.deliveryChannel = deliveryChannel
    }

    /// Capture a verification photo and deliver it to the companion app.
    /// Returns nil if camera is unavailable (graceful degradation per FR-D04d).
    func verify(
        adherenceLogId: UUID,
        targetDeviceId: String
    ) async -> PhotoVerificationStatus {
        let captureResult = await captureSession.captureStillImage()

        switch captureResult {
        case .cameraUnavailable(let reason):
            print("[PhotoVerifier] Camera unavailable: \(reason)")
            return .unavailable

        case .captureFailed(let reason):
            print("[PhotoVerifier] Capture failed: \(reason)")
            return .unavailable

        case .captured(let imageData):
            return await processAndDeliver(
                imageData: imageData,
                adherenceLogId: adherenceLogId,
                targetDeviceId: targetDeviceId
            )
        }
    }

    // MARK: - Private

    private func processAndDeliver(
        imageData: Data,
        adherenceLogId: UUID,
        targetDeviceId: String
    ) async -> PhotoVerificationStatus {
        // Compress to max 200 KB if needed
        let compressed = compressIfNeeded(imageData)

        let photo = VerificationPhoto(
            id: UUID(),
            adherenceLogId: adherenceLogId,
            capturedAt: Date(),
            deliveredAt: nil,
            deletedFromDeviceAt: nil,
            imageData: compressed
        )

        let deliveryResult = await deliveryChannel.deliver(photo: photo, to: targetDeviceId)

        switch deliveryResult {
        case .delivered:
            // Photo will be deleted from primary device within 1 hour (FR-D04b)
            // The deletion is handled by a separate cleanup task
            return .delivered

        case .deliveryFailed:
            // Retain photo on device for retry
            return .captured
        }
    }

    private func compressIfNeeded(_ imageData: Data) -> Data {
        if imageData.count <= maxImageSizeBytes {
            return imageData
        }

        // Simple compression: reduce quality iteratively
        // In production: use platform image APIs (UIImage JPEG compression on iOS,
        // Bitmap compress on Android)
        var quality: CGFloat = 0.9
        var compressed = imageData

        while compressed.count > maxImageSizeBytes && quality > 0.1 {
            quality -= 0.1
            // Placeholder: in production, call UIImage(data:)?.jpegData(compressionQuality: quality)
            compressed = Data(imageData.prefix(Int(Double(imageData.count) * quality)))
        }

        return Data(compressed.prefix(maxImageSizeBytes))
    }
}

File: android/app/src/main/java/com/elderlyassistant/services/medication/PhotoVerifier.kt
package com.elderlyassistant.services.medication

import java.util.UUID

// MARK: - Photo Verifier (FR-D04a through FR-D04e)

sealed class PhotoCaptureResult {
    data class Captured(val imageData: ByteArray) : PhotoCaptureResult()
    data class CameraUnavailable(val reason: String) : PhotoCaptureResult()
    data class CaptureFailed(val reason: String) : PhotoCaptureResult()
}

sealed class PhotoDeliveryResult {
    data class Delivered(val deliveredAt: Long) : PhotoDeliveryResult()
    data class DeliveryFailed(val reason: String) : PhotoDeliveryResult()
}

interface PhotoCaptureSession {
    suspend fun captureStillImage(): PhotoCaptureResult
}

interface PhotoDeliveryChannel {
    suspend fun deliver(photo: VerificationPhoto, targetDeviceId: String): PhotoDeliveryResult
}

class PhotoVerifier(
    private val captureSession: PhotoCaptureSession,
    private val deliveryChannel: PhotoDeliveryChannel
) {
    companion object {
        private const val MAX_IMAGE_SIZE_BYTES = 200 * 1024  // 200 KB per FR-D04b
    }

    suspend fun verify(
        adherenceLogId: UUID,
        targetDeviceId: String
    ): PhotoVerificationStatus {
        return when (val captureResult = captureSession.captureStillImage()) {
            is PhotoCaptureResult.CameraUnavailable -> {
                println("[PhotoVerifier] Camera unavailable: ${captureResult.reason}")
                PhotoVerificationStatus.UNAVAILABLE
            }
            is PhotoCaptureResult.CaptureFailed -> {
                println("[PhotoVerifier] Capture failed: ${captureResult.reason}")
                PhotoVerificationStatus.UNAVAILABLE
            }
            is PhotoCaptureResult.Captured ->
                processAndDeliver(captureResult.imageData, adherenceLogId, targetDeviceId)
        }
    }

    private suspend fun processAndDeliver(
        imageData: ByteArray,
        adherenceLogId: UUID,
        targetDeviceId: String
    ): PhotoVerificationStatus {
        val compressed = compressIfNeeded(imageData)

        val photo = VerificationPhoto(
            id = UUID.randomUUID(),
            adherenceLogId = adherenceLogId,
            capturedAt = System.currentTimeMillis(),
            deliveredAt = null,
            deletedFromDeviceAt = null,
            imageData = compressed
        )

        return when (val result = deliveryChannel.deliver(photo, targetDeviceId)) {
            is PhotoDeliveryResult.Delivered -> PhotoVerificationStatus.DELIVERED
            is PhotoDeliveryResult.DeliveryFailed -> PhotoVerificationStatus.CAPTURED
        }
    }

    private fun compressIfNeeded(imageData: ByteArray): ByteArray {
        if (imageData.size <= MAX_IMAGE_SIZE_BYTES) return imageData

        var quality = 90
        var compressed = imageData

        while (compressed.size > MAX_IMAGE_SIZE_BYTES && quality > 10) {
            quality -= 10
            // In production: use Bitmap.compress(Bitmap.CompressFormat.JPEG, quality, ...)
            compressed = compressed.take((imageData.size * quality) / 100).toByteArray()
        }

        return compressed.take(MAX_IMAGE_SIZE_BYTES).toByteArray()
    }
}
